ares/n64/pi/bus.hpp

```diff
@@ -0,0 +1,79 @@
+template<u32 Size>
+inline auto PI::busRead(u32 address) -> u64 {
+  static constexpr u64 unmapped = 0;
+
+  if (unlikely(io.ioBusy)) {
+    io.ioBusy = 0; //technically, we should wait until Queue::PI_BUS_Write
+    if constexpr(Size == Byte) {
+      return io.busLatch >> ((~address & 3) * 8);
+    }
+    if constexpr(Size == Half) {
+      return io.busLatch >> ((~address & 1) * 16);
+    }
+    if constexpr(Size == Word) {
+      return io.busLatch;
+    }
+    if constexpr(Size == Dual) {
+      return io.busLatch;
+    }
+  }
+
+  if(address <= 0x0500'03ff) return dd.c2s.read<Size>(address);
+  if(address <= 0x0500'04ff) return dd.ds.read<Size>(address);
+  if(address <= 0x0500'057f) return dd.read<Size>(address);
+  if(address <= 0x0500'05bf) return dd.ms.read<Size>(address);
+  if(address <= 0x05ff'ffff) return unmapped;
+  if(address <= 0x063f'ffff) return dd.iplrom.read<Size>(address);
+  if(address <= 0x07ff'ffff) return unmapped;
+  if(address <= 0x0fff'ffff) {
+    if(cartridge.ram  ) return cartridge.ram.read<Size>(address);
+    if(cartridge.flash) return cartridge.flash.read<Size>(address);
+    return unmapped;
+  }
+  if(address >= 0x13ff'0000 && address <= 0x13ff'ffff) {
+    return cartridge.isviewer.read<Size>(address);
+  }
+  return cartridge.rom.read<Size>(address);
+}
+
+template<u32 Size>
+inline auto PI::busWrite(u32 address, u64 data) -> void {
+  if(io.ioBusy) return;
```

Comment:
How will this interact with PI::dmaRead()? It makes back-to-back writes in a loop.

Reply:
Right, that is broken now. Suggestions?

We could have an internal function that does performs a word write (eg busWriteWord) without any PI busy concern, and call it from both dmaWrite and busWrite.

Reply:
I went ahead and did this. Now there is a lower-level layer of functions called busWrite/busRead that only can do 16-bit and 32-bit accesses (mimicking what happens on the PI bus: DMA can do both 16-bit and 32-bit accesses while CPU will always do 32-bit accesses).

These are now called from readWord / writeWord so that the special word mangling performed by Memory::IO get applied also to bus reads/writes and the code doesn't need to be duplicated like in the previous version of this PR.

Given this new approach, the quirky behavior of cartridge 8-bit/16-bit reads is now more straightforward and simply happens because ROM is a physical 16-bit device so the 32-bit access from PI is correctly aligned to 16-bit. This is emulated via a new class called Memory::Readable16.